lib/recursive-methods.rb

```diff
@@ -1,49 +1,148 @@
+
+  return reverse(s[1..-1]) + s[0]    # Solution (1)
+  # return s[-1] + reverse(s[0..-2]) # Solution (2)
+end
+
+
+
+# Time complexity: O(n) => where n is the length of s
+# Space complexity: O(n) => because of the System Stack
+def reverse_inplace(s, i = 0, j = s.length - 1)
+  return s if i >= j
+
+  s[i], s[j] = s[j], s[i]
+
+  return reverse_inplace(s, i + 1, j - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+
+
+# Time complexity: O(n) => n depends on the input size
+# Space complexity: O(n) => because of the System Stack
+def bunny(n) 
+  return 0 if n == 0
+  return 2 + bunny(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+
+
+
+# Solution (1)
+# Time complexity: O(n) => where n is the length of s
+# Space complexity: O(n) => because of the System Stack
+def nested(s, low = 0, high = s.length - 1)
+  if s.empty? || low >= high
+    return true 
+  elsif s.length.odd?
+    return false 
+  end 
+
+  return (s[low] == "(" && s[high] == ")") && nested(s, low + 1, high - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def nested(s)
-    raise NotImplementedError, "Method not implemented"
+
+# # Solution (2)
+# # Time complexity: O(n) => where n is the length of s
+# # Space complexity: O(n^2) => because it creates a new string for each recursion
+# def nested(s)
+#   return true if s.empty?
+#   return false if s.length.odd?
+
+#   return (s[0] == "(" && s[-1] == ")") && nested(s[1..-2])
+# end
+
+
+# Time complexity: O(n) => where n is the length of array
+# Space complexity: O(n) => because of the System Stack
+def search(array, value, i = 0)
+  return false if array.empty? || i >= array.length
+  return true if array[i] == value
+
+  return search(array, value, i + 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def search(array, value)
-    raise NotImplementedError, "Method not implemented"
+
+# Solution (1)
+# Time complexity: O(n) => where n is the length of s
+# Space complexity: O(n) => Because of the System Stack
+def is_palindrome(s, low = 0, high = s.length - 1)
+  if s.length < 2 || low >= high 
+    return true
+  elsif s[low] != s[high]
+    return false 
+  else  
+    return is_palindrome(s, low + 1, high - 1)
+  end 
+
+  return true
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def is_palindrome(s)
-    raise NotImplementedError, "Method not implemented"
+
+# # Solution (2)
+# # Time complexity: O(n) => where n is the length of s
+# # Space complexity: O(n^2) => because it creates a new string (s[1..-2]) for each recursion
+# def is_palindrome(s)
+#   return true if s.length < 2
+#   return false if s[0] != s[-1]
+
+#   return is_palindrome(s[1..-2])
+# end
+
+
+
+# Time complexity: O(n) => (if I consider that n < m)
+# Space complexity: O(n) => because of the System Stack
+def digit_match(n, m, sum = 0)
```

Comment:
The Big-O here for time complexity depends on what you define `n` as.  If `n` is the number of digits in the smaller of the two parameters then the algorithm is O(n).  If you set `n` to be the value of the to smaller of the two numbers then it's O(log10 n).